src/web-extensions.c

```diff
@@ -0,0 +1,861 @@
+    }
+
+  host_path = find_messaging_host (web_extensions_session->mode,
+                                   arg_name, arg_extension_or_origin,
+                                   NULL, NULL, &json_manifest, &error);
+  if (!host_path)
+    {
+      g_dbus_method_invocation_return_gerror (invocation, error);
+      return TRUE;
+    }
+
+  xdp_dbus_web_extensions_complete_get_manifest (object, invocation, json_manifest);
+  return TRUE;
+}
+
+static void
+handle_start_in_thread (GTask *task,
+                        gpointer source_object,
+                        gpointer task_data,
+                        GCancellable *cancellable)
+{
+  Request *request = (Request *)task_data;
+  Session *session;
+  WebExtensionsSession *web_extensions_session;
+  const char *arg_name;
+  char *arg_extension_or_origin;
+  const char *app_id;
+  g_autofree char *host_path = NULL;
+  g_autofree char *description = NULL;
+  g_autofree char *manifest_filename = NULL;
+  guint response = XDG_DESKTOP_PORTAL_RESPONSE_OTHER;
+  gboolean should_close_session;
+  Permission permission;
+  gboolean allowed;
+  char *argv[] = {NULL, NULL, NULL, NULL};
+  g_autoptr(GError) error = NULL;
+
+  REQUEST_AUTOLOCK (request);
+  session = g_object_get_data (G_OBJECT (request), "session");
+  SESSION_AUTOLOCK_UNREF (g_object_ref (session));
+  g_object_set_data (G_OBJECT (request), "session", NULL);
+  web_extensions_session = (WebExtensionsSession *)session;
+
+  if (!request->exported || web_extensions_session->state != WEB_EXTENSIONS_SESSION_STATE_STARTING)
+    goto out;
+
+  arg_name = g_object_get_data (G_OBJECT (request), "name");
+  arg_extension_or_origin = g_object_get_data (G_OBJECT (request), "extension-or-origin");
+
+  host_path = find_messaging_host (web_extensions_session->mode,
+                                   arg_name, arg_extension_or_origin,
+                                   &description, &manifest_filename, NULL,
+                                   &error);
+  if (host_path == NULL)
+    {
+      g_warning ("Could not find WebExtensions backend: %s", error->message);
+      fflush(stderr);
+      fflush(stdout);
+      goto out;
+    }
+
+  app_id = xdp_app_info_get_id (request->app_info);
+  permission = get_permission_sync (app_id, PERMISSION_TABLE, arg_name);
+  if (permission == PERMISSION_ASK || permission == PERMISSION_UNSET)
+    {
+      guint access_response = 2;
+      g_autoptr(GVariant) access_results = NULL;
+      GVariantBuilder opt_builder;
+      g_autoptr(GAppInfo) info = NULL;
+      const char *display_name;
+      g_autofree gchar *app_info_id = NULL;
+      g_autofree gchar *title = NULL;
+      g_autofree gchar *subtitle = NULL;
+      g_autofree gchar *body = NULL;
+
+      info = xdp_app_info_load_app_info (request->app_info);
+      if (info)
+        {
+          g_auto(GStrv) app_id_components = g_strsplit (g_app_info_get_id (info), ".desktop", 2);
+          app_info_id = g_strdup (app_id_components[0]);
```

Comment:
According to the [documentation of `g_app_info_get_id`](https://docs.gtk.org/gio/method.AppInfo.get_id.html), its return value can be NULL. Is it possible for this to be NULL in the context of this function? If yes, a NULL check may be needed:

```suggestion
          const char *raw_id = g_app_info_get_id (info);
          if (raw_id) {
            g_auto(GStrv) app_id_components = g_strsplit (raw_id, ".desktop", 2);
            app_info_id = g_strdup (app_id_components[0]);
          }
```